### ISSUE TITLE
Simplify NuttX initialization

### ISSUE THREAD

**Issue by acassis** [apache/nuttx#11321]
Currently NuttX has many initialization options each one used for some specific case, but at least CONFIG_NSH_ARCHINIT that is used to do drivers initialization when NSH is enabled could be removed in favor of some generic option like BOARD_LATE_INITIALIZATION that will work for all applications (inclusive to NSH itself).

I raised this question some time ago here: https://github.com/apache/nuttx/issues/10821 but didn't heard anything from my peers.

But now that @0528Mike came with this PR: https://github.com/apache/nuttx/pull/11306 we could fix this issue merging CONFIG_NSH_ARCHINIT and BOARD_LATE_INITIALIZATION to avoid issues like this faced from a NuttX user:  https://github.com/apache/nuttx/issues/10821 and for many people that want to user their application as ENTRYPOINT on NuttX.

**Comment by xiaoxiang781216** [comment]
Yes, I think the suggestion is reasonable.

**Comment by raiden00pl** [comment]
+1 from me

**Comment by raiden00pl** [comment]
hi, are we trying to solve this problem? Does anyone have any objections to removing `CONFIG_NSH_ARCHINIT` ?
This option is slowly starting to cause problems in the case of PCI devices. Some OS components that have a PCI based virtio backend need to be initialized in the correct order (e.g. coredump with PCI based virtblk device), which is impossible to achieve with `CONFIG_NSH_ARCHINIT`.

**Comment by xiaoxiang781216** [comment]
NSH_ARCHINIT can be achieved in board initializiton code, so it could be removed to avoid the confusion.

**Comment by gregory-nutt** [comment]
> NSH_ARCHINIT can be achieved in board initializiton code, so it could be removed to avoid the confusion.

Agreed.  But removing CONFIG_NSH_ARCHINIT will be a very big job because it is used by 1246 board configurations.

```
$ find boards -name defconfig | xargs grep CONFIG_NSH_ARCHINIT=y | wc -
   1246    1246  100027 
```
That would be a LOT of breakage!

**Comment by raiden00pl** [comment]
Yes, it seems like a lot of work. But it will only get worse over time as new boards come in.

We can automate most of the work. What we need to do is:

1. remove all `CONFIG_NSH_ARCHINIT` from configs 
2. add `CONFIG_BOARD_LATE_INITIALIZATION` to all modified configs
3. clean up after removing `CONFIG_NSH_ARCHINIT`.

Is there anything more to do ?

The last step most likely must be done manually. But there are "only" 115 files that need modification:

```
$ find . -type f ! -name "defconfig" -exec grep -l "CONFIG_NSH_ARCHINIT" {} + | wc                                                                                                                              
    115     115    6899
```

**Comment by gregory-nutt** [comment]
> Is there anything more to do ?

I think a proper solution would be the complete remove of the archi-init logic, not just disabling it.  I think that would cause a lot of confusion in the future.

This would require removal the the BOARD_CTRL log, removal of the arch-init entry point to the board initialization logic, and removal of all calls to the removed BOARD_CTRL logic.

**Comment by raiden00pl** [comment]
It makes sense. Such a change deserve major version increase during release, as this is substantial change for users. Probably this issue is a good milestone for NuttX 13.0.0 release

**Comment by linguini1** [comment]
I know that this PR was closed, and it appears there are no plans for the author to continue it, but is there a general agreement that something similar to this init approach would be the preferred change? https://github.com/apache/nuttx-apps/pull/3192

I am envisioning some pure initializer (like nxinit was) that allows the user to choose 1+ different processes to start after boot time, and also which takes care of some common logic like starting up USB, etc.

Once such an initializer is accepted by the community, I think we can then move to transform NSH from this "do-all" solution to a pure shell that users can choose to include or not.

**Comment by raiden00pl** [comment]
this issue is about too complex board-specific initialization process on the kernel side, it has nothing to do with apps and it can't be solved by the mentioned PR

**Comment by linguini1** [comment]
Hi, I'm interested in working on this. To clarify what the desired solution is, we want to remove `NSH_ARCHINIT` in apps and replace it with `CONFIG_BOARD_LATE_INITIALIZE` in all board configurations that are currently relying on it? And then, in addition, we wish to completely remove the `boardctl` logic for `BOARDIOC_INIT` in both kernel and apps in favour of just having the user enabled `CONFIG_BOARD_LATE_INITIALIZE`?

So, this involves:
* Removing all of the `NSH_ARCHINIT` logic
* Replacing `CONFIG_NSH_ARCHINIT` with `CONFIG_BOARD_LATE_INITIALIZATION` in all upstream defconfig files
* Verifying that all logic in `board_late_initialize` is equivalent to the logic performed by `NSH_ARCHINIT` for each modified board
* Removal of the `boardctl(BOARDIOC_INIT, 0)` logic for _all_ boards in favour of the late initialization path (i.e. no more `board_app_initialize`)?
* Removal of the `boardctl(BOARDIOC_INIT, 0)` logic present in _all_ apps

**Comment by raiden00pl** [comment]
@linguini1 correct. One more step will be to clean `boards/` from references to the removed logic and this is probably where the most work is:

https://github.com/apache/nuttx/blob/56112efbd3f5027dda1cf8498c94f8e3b311e80a/boards/arm/stm32/stm32f4discovery/src/stm32_appinit.c#L67

https://github.com/apache/nuttx/blob/56112efbd3f5027dda1cf8498c94f8e3b311e80a/boards/arm/stm32/stm32f4discovery/src/CMakeLists.txt#L96

https://github.com/apache/nuttx/blob/56112efbd3f5027dda1cf8498c94f8e3b311e80a/boards/arm/stm32/stm32f4discovery/src/Make.defs#L102

https://github.com/apache/nuttx/blob/56112efbd3f5027dda1cf8498c94f8e3b311e80a/boards/arm/stm32/stm32f4discovery/src/stm32_bringup.c#L212-L217

for stm32 chips this is rather standardized, I'm not sure about other architectures

**Comment by linguini1** [comment]
Okay that makes sense. As a follow-up question: it seems that basically all of the applications are either relying on `boardctl` to do initialization or NSH to archinit. If we're removing archinit and the boardctl stuff in favour of `CONFIG_BOARD_LATE_INITIALIZE`, I suppose it would be advantageous for me just to change the default to `CONFIG_BOARD_LATE_INITIALIZE=y` in Kconfig? I don't really see any scenario where a user wouldn't want to have board initialization performed seeing as there will no longer be any other avenue for it to occur?

**Comment by raiden00pl** [comment]
@linguini1 Some applications with very low-power requirements may benefit from not having board init executed until absolutely necessary. The user can set `CONFIG_BOARD_LATE_INITIALIZE=n` and initialize the peripherals using `BOARDIOC_FINALINIT`. 

Let's say:

0. device boot, all board peripherals in reset state and we have to minimize CPU operations, 
1. an external event occurs and device has to do some work,
2. the device initializes board peripherals required for work,
3. device does some work,
4. device resets, starting the cycle from the beginning. 

This seems to me to be the only practical use for not initializing board peripherals at boot. By default, peripherals should be disabled after initialization, but it is still possible that initialization itself may increase current consumption in some extreme cases.

**Comment by linguini1** [comment]
Some boards don't appear to have a `board_late_initialize` defined, only the app_init logic. For those my current plan is to take the contents of the `board_app_initialize` function and put it in a new `board_late_initialize` function.

**Comment by linguini1** [comment]
Another note to possibly re-visit: some boards have their late init logic in `*_boot.c` and others in `*_boardinit.c`.

**Comment by acassis** [comment]
@linguini1 is the boot process is a mess, there are originally the xxx_boot.c and later xxx_bringup.c. Later xxx_appinit.c and finally there is only one xxx_init.c: avr/avrdx/breadxavr/src/avrdx_init.c

I think we need some consensus before changing it! For visibility: @raiden00pl @simbit18 @hartmannathan @xiaoxiang781216 @lupyuen  @jerpelea @cederom (sorry if I missed another people interested on it)

**Comment by cederom** [comment]
Before we remove anything it would be good to review current boot architecture and boot sequence and then think on how things can be improved at the architecture level first? :-)

https://nuttx.apache.org/docs/latest/guides/port_bootsequence.html

**Comment by linguini1** [comment]
@acassis @cederom 
I don't plan to change anything like that as part of this issue (renaming the files and making some standard set of initialization code, etc., is beyond the scope). For now all I am doing is removing the NSH_ARCHINIT and `board_app_initialize` logic in favour of `BOARD_LATE_INITIALIZE` as was agreed upon in this issue. After the fact, if we want to make more improvements to the boot process, I think that can be part of a different discussion and issue tracker!

**Comment by lupyuen** [comment]
FYI I have a couple of articles explaining the NuttX Boot Flow...

- [__NuttX Boot Flow in PDF__](https://github.com/lupyuen/nuttx-arm64-emulator/blob/avaota/nuttx-boot-flow.pdf) / [__SVG__](https://github.com/lupyuen/nuttx-arm64-emulator/blob/avaota/nuttx-boot-flow.svg) / [__PNG__](https://github.com/lupyuen/nuttx-arm64-emulator/blob/avaota/nuttx-boot-flow.png)

- Based on the [__Mermaid Flowchat__](https://raw.githubusercontent.com/lupyuen/nuttx-arm64-emulator/refs/heads/avaota/nuttx-boot-flow.mmd)

- Which was generated by our [Arm64 Avaota-A1 Emulator](https://lupyuen.org/articles/unicorn4)

We also fed the above [__Mermaid Flowchat__](https://raw.githubusercontent.com/lupyuen/nuttx-arm64-emulator/refs/heads/avaota/nuttx-boot-flow.mmd) to Gemini LLM to get these...

- [__LLM explains NuttX Boot Flow (Short Version)__](https://gist.github.com/lupyuen/b7d937c302d1926f62cea3411ca0b3c6)

- [__LLM explains NuttX Boot Flow (Long Version)__](https://docs.google.com/document/d/1qYkBu3ca3o5BXdwtUpe0EirMv9PpMOdmf7QBnqGFJkA/edit?tab=t.0)

See also:

- [NuttX Boot Flow for RISC-V Ox64 BL808](https://lupyuen.org/articles/ox2#appendix-nuttx-boot-flow)

- [NuttX Boot Flow for Arm64 PinePhone](https://lupyuen.org/articles/unicorn2#pinephone-boots-nuttx)

**Comment by hartmannathan** [comment]
> @linguini1 is the boot process is a mess, there are originally the xxx_boot.c and later xxx_bringup.c. Later xxx_appinit.c and finally there is only one xxx_init.c: avr/avrdx/breadxavr/src/avrdx_init.c
> 
> I think we need some consensus before changing it! For visibility: @raiden00pl @simbit18 @hartmannathan @xiaoxiang781216 @lupyuen  @jerpelea @cederom (sorry if I missed another people interested on it)

Hi all,

I'm in favor of improving the boot process. Just please remember that 3rd party developers may have private, out-of-tree boards that utilize some of these functions in their initialization. (My boards use board_late_initialize, but others may have done differently.) So whatever approach we take to simplify and clean up the boot process, we need to make sure that developers can have the hooks they need to initialize their stuff properly. And we need to make sure that any changes people will need to make, will be properly documented in the release notes!!!

**Comment by linguini1** [comment]
> I'm in favor of improving the boot process. Just please remember that 3rd party developers may have private, out-of-tree boards that utilize some of these functions in their initialization. (My boards use board_late_initialize, but others may have done differently.) So whatever approach we take to simplify and clean up the boot process, we need to make sure that developers can have the hooks they need to initialize their stuff properly. And we need to make sure that any changes people will need to make, will be properly documented in the release notes!!!

I agree. Right now the idea is to gut the BOARDIOC_INIT and NSH_ARCHINIT stuff.
All users should then be able to choose between two options:

1. Just enable CONFIG_BOARD_LATE_INITIALIZE and let the NuttX kernel handle the
initialization (the elegant solution)

2. For cases like raiden pointed out where app developers _really_ need
userspace control over initialization, disable CONFIG_BOARD_LATE_INITIALIZE and
implement a custom `board_finalinitialization` to call from userspace via
`boardctl(BOARDIOC_FINALINIT)`.

As @cederom suggested (I think somewhere on the mailing list) I also plan to include a "quick fix" message in both the PR descriptions and commit messages that indicate to the end user what they need to change in their defconfig/custom board logic/custom app to fix any breakages.

**Comment by cederom** [comment]
@patacongo pointed to older documentation [1] that we may use as reference point, and probably we should update our current documentation :-)

I like the clear distinction:
1. Phase A - The hardware-specific power-on reset initialization (ARCH_INIT / INIT_ARCH?)
2. Phase B - NuttX RTOS initialization (NX_INIT / INIT_NX?)
3. Phase C - Application Initialization (APP_INIT / INIT_APP? nsh init here?)

But there is no info about where and how board is initialized (BOARD_INIT / INIT_BOARD?) maybe we could add one (as C before APP, then APP becomes D)? :-)

@linguini1 added plantuml support for documentation so we can now auto-generate awesome diagrams from simple text :-)

[1] https://cwiki.apache.org/confluence/display/NUTTX/NuttX+Initialization+Sequence

**Comment by acassis** [comment]
> > [@linguini1](https://github.com/linguini1) is the boot process is a mess, there are originally the xxx_boot.c and later xxx_bringup.c. Later xxx_appinit.c and finally there is only one xxx_init.c: avr/avrdx/breadxavr/src/avrdx_init.c
> > I think we need some consensus before changing it! For visibility: [@raiden00pl](https://github.com/raiden00pl) [@simbit18](https://github.com/simbit18) [@hartmannathan](https://github.com/hartmannathan) [@xiaoxiang781216](https://github.com/xiaoxiang781216) [@lupyuen](https://github.com/lupyuen)  [@jerpelea](https://github.com/jerpelea) [@cederom](https://github.com/cederom) (sorry if I missed another people interested on it)
> 
> Hi all,
> 
> I'm in favor of improving the boot process. Just please remember that 3rd party developers may have private, out-of-tree boards that utilize some of these functions in their initialization. (My boards use board_late_initialize, but others may have done differently.) So whatever approach we take to simplify and clean up the boot process, we need to make sure that developers can have the hooks they need to initialize their stuff properly. And we need to make sure that any changes people will need to make, will be properly documented in the release notes!!!

@hartmannathan I agree that we need to worry about how people are keeping theirs out of tree boards, but instead of creating hooks, it is better to Document how to change to use board_late_initialize or the other alternatives. If I'm not wrong, there is already a "hook" for board early initialize. If we keep many hooks in the kernel, the chaos will continue. The board initialization on NuttX is already very complex, we need to simplify keeping the most common use cases.

**Comment by cederom** [comment]
Ugh, I just realized this is already on the 13's list :-P

**Comment by jerpelea** [comment]
> Ugh, I just realized this is already on the 13's list :-P

@cederom 13.0.0 will be branched out on 1 March

**Comment by acassis** [comment]
> > Ugh, I just realized this is already on the 13's list :-P
> 
> [@cederom](https://github.com/cederom) 13.0.0 will be branched out on 1 March

Alin, let's use it as unstable/testing release only, NuttX 13 will not help to promote NuttX :-D

**Comment by cederom** [comment]
> > > @cederom: Ugh, I just realized this is already on the 13's list :-P
> > 
> > 
> > @jerpelea: [@cederom](https://github.com/cederom) 13.0.0 will be branched out on 1 March
> 
> @acassis: Alin, let's use it as unstable/testing release only, NuttX 13 will not help to promote NuttX :-D

Yeah I would stick to 12 too until all tedious work is done and we are really proud to announce 13 :-)

**Comment by acassis** [comment]
@xiaoxiang781216 Ok, let's open a new issue to address the NX Init missing part
